File: 01--calorie-counting/b.hs
-- Usage:
--   runhaskell b.hs < in

import Data.List (sort)
import Data.Function ((&))

main = interact solve

solve text =
  text
  & paragraphs
  & map
    (
      \paragraph ->
        paragraph
        & lines
        & map
          (
            \line ->
              line
              & (read :: (String -> Int))
          )
        & sum
    )
  & sort
  & reverse
  & take 3
  & sum
  & show

paragraphs text = splitWithStr "\n\n" text

-- https://stackoverflow.com/a/10967044
splitWithStr :: Eq a => [a] -> [a] -> [[a]]
splitWithStr x y = func x y [[]]
  where
    func x [] z = reverse $ map (reverse) z
    func x (y:ys) (z:zs) = if (take (length x) (y:ys)) == x then
      func x (drop (length x) (y:ys)) ([]:(z:zs))
    else
      func x ys ((y:z):zs)

File: 01--calorie-counting/b.js
const { readFileSync } = require('fs');

const answer = readFileSync('in', 'utf-8')
  .trimEnd()
  .split('\n\n')
  .map(paragraph =>
    paragraph
      .split('\n')
      .map(line => +line)
      .reduce((a, b) => a + b) // sum
  )
  .sort((a, b) => b - a) // sort descending
  .slice(0, 3)
  .reduce((a, b) => a + b) // sum

console.log(answer);
